File: concepts/06_text_binary_and_serialization/03_regular_expressions_and_state/python/test_02_zero_width_and_failure_boundaries.py
"""零宽匹配、游标推进与编译失败。

共同问题：零宽成功如何避免无限迭代；复用 matcher 是否泄漏游标；
非法模式在什么阶段报告。
"""

# polyglot-family: text_binary_and_serialization
# polyglot-concept: regular_expressions_and_state
# polyglot-related: languages/python/stdlib/036-040_text_processing/test_036_regular_expressions.py

import re

import pytest


def test_finditer_advances_after_zero_width_matches():
    pattern = re.compile(r"(?=.)")

    matches = list(pattern.finditer("ab"))

    assert [match.span() for match in matches] == [(0, 0), (1, 1)]
    assert [match.group() for match in matches] == ["", ""]


def test_position_is_an_explicit_call_argument_not_pattern_state():
    pattern = re.compile(r"\d")
    text = "1a2"

    assert pattern.search(text, pos=0).span() == (0, 1)
    assert pattern.search(text, pos=1).span() == (2, 3)
    assert pattern.search(text, pos=0).span() == (0, 1)


def test_invalid_pattern_fails_during_compile():
    with pytest.raises(re.error):
        re.compile("(")

File: concepts/06_text_binary_and_serialization/03_regular_expressions_and_state/r/test_02_zero_width_and_failure_boundaries.R
# polyglot-family: text_binary_and_serialization
# polyglot-concept: regular_expressions_and_state
# polyglot-related: languages/r/standard_library/13_io_text_time_and_numeric_workflows/
# polyglot-related+: test_099_csv_encodings_locale_and_regular_expressions.R
#
# 共同问题：零宽匹配怎样前进，非法模式和无匹配怎样表达。
# 对照观察：`gregexpr` 返回位置向量；无匹配用 -1，非法 PCRE 模式发出 error。

zero_width <- gregexpr("(?=a)", "aa", perl = TRUE)[[1L]]
missing <- regexpr("z", "aa", fixed = TRUE)
invalid <- tryCatch(grepl("(", "text", perl = TRUE), error = identity)

stopifnot(
    identical(as.integer(zero_width), c(1L, 2L)),
    identical(as.integer(missing), -1L),
    inherits(invalid, "error")
)
